crates/sui-config/src/genesis.rs

```diff
@@ -691,25 +751,231 @@ impl Builder {
             objects,
         };
 
-        // Verify that all the validators were properly created onchain
-        let system_object = genesis
+        // Verify that all on-chain state was properly created
+        self.validate().unwrap();
+
+        genesis
+    }
+
+    pub fn validate(&self) -> Result<(), anyhow::Error> {
+        if !self.parameters.allow_insertion_of_extra_objects && !self.objects.is_empty() {
+            bail!("extra objects are disallowed");
+        }
+
+        let GenesisChainParameters {
+            protocol_version,
+            chain_start_timestamp_ms,
+            epoch_duration_ms,
+            stake_subsidy_start_epoch,
+            stake_subsidy_initial_distribution_amount,
+            stake_subsidy_period_length,
+            stake_subsidy_decrease_rate,
+            max_validator_count,
+            min_validator_joining_stake,
+            validator_low_stake_threshold,
+            validator_very_low_stake_threshold,
+            validator_low_stake_grace_period,
+        } = self.parameters.to_genesis_chain_parameters();
+
+        for validator in self.validators.values() {
+            validator.validate().with_context(|| {
+                format!(
+                    "metadata for validator {} is invalid",
+                    validator.info.name()
+                )
+            })?;
+        }
+
+        if let Some(token_distribution_schedule) = &self.token_distribution_schedule {
+            token_distribution_schedule.validate();
+            token_distribution_schedule.check_all_stake_operations_are_for_valid_validators(
+                self.validators.values().map(|v| v.info.sui_address()),
+            );
+        }
+
+        let Some(unsigned_genesis) = self.unsigned_genesis_checkpoint() else {
```

Comment:
Slightly confused about what's happening here. It seems that you use this `validate` function in two places? Before we build the unsigned genesis and after. In the before case we only perform the checks above line 796 to check validator metadata validity and that token distribution schedule is valid. In the after case we additionally perform a lot of checks making sure that what's onchain in 0x5 matches our expectation. Is that right? Would it be better if we separate this into two functions `validate_genesis_input` and `validate_genesis_output`? The checks before line 796 belong to the first function and the checks after line 796 the second.

Reply:
Yeah this is more about validating the state contained in the Builder which is why it has this split. I can split it, but i'll probably end up calling both functions in every place as well.